x/bsonx/bsoncore/bsonbuilder.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+// Package bsoncore contains functions that can be used to encode and decode bson
+// elements and values to or from a slice of bytes. These functions are aimed at
+// allowing low level manipulation of bson and can be used to build a higher
+// level bson library.
+package bsoncore
+
+import (
+	"strconv"
+
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+)
+
+// DocumentBuilder builds a bson document
+type DocumentBuilder struct {
+	doc   []byte
+	index []int32
+}
+
+// ArrayBuilder builds a bson array
+type ArrayBuilder struct {
+	arr   []byte
+	index []int32
+	key   []int
+}
+
+// startDocument reserves the document's length and set the index to where the length begins
+func (b *DocumentBuilder) startDocument() *DocumentBuilder {
+	var index int32
+	index, b.doc = AppendDocumentStart(b.doc)
+	b.index = append(b.index, index)
+	return b
+}
+
+// startArray reserves the array's length and sets the index to where the length begins
+func (a *ArrayBuilder) startArray() *ArrayBuilder {
+	var index int32
+	index, a.arr = AppendArrayStart(a.arr)
+	a.index = append(a.index, index)
+	a.key = append(a.key, 0)
+	return a
+}
+
+// NewDocumentBuilder creates a new DocumentBuilder
+func NewDocumentBuilder() *DocumentBuilder {
+	b := (&DocumentBuilder{}).startDocument()
+	return b
+}
+
+// NewArrayBuilder creates a new ArrayBuilder
+func NewArrayBuilder() *ArrayBuilder {
+	a := (&ArrayBuilder{}).startArray()
+	return a
+}
+
+// Build updates the length of the document and index to the beginning of the documents length
+// bytes, then returns the document (bson bytes)
+func (b *DocumentBuilder) Build() []byte {
```

Comment:
We should return `bsoncore.Document` rather than `[]byte`. This has the advantage of some type safety and `bsoncore.Document` has a custom string function that allows it to print out as extended JSON, so it's easier to debug.

Reply:
done.